[Incubator/kafka] Topic creation and configuration

t-d-d opened the pull request:
Also:
- Bump cp-kafka to 4.1.1
- Template filename changes for consistency
- Fix  jmx sidecar to swallow errors instead of bringing down the pod (which happens with slow broker startup.)

benjigoldberg commented:
/assign @benjigoldberg

benjigoldberg commented:
/ok-to-test

benjigoldberg commented:
@t-d-d pretty nice functionality in this PR! The changes LGTM so far, but I havent had a chance to toy with it yet. Will try and do it this afternoon.

t-d-d committed:
[Incubator/kafka] Topic creation and configuration

Also:
-Bump cp-kafka to 4.1.1
-Template filename changes for consistency
-Fix jmx sidecar to swallow errors instead of bringing down the pod (which happens with slow broker startup.)

benjigoldberg commented:
@t-d-d  ok -- I've finally found the time to work with this a bit and I think this looks good to me. If you resolve the `Chart.yaml` conflict I think we're good to go

benjigoldberg commented on incubator/kafka/templates/configmap-config.yaml:
```diff
@@ -0,0 +1,38 @@
+    kafka-topics --zookeeper {{ $zk }} --create --if-not-exists --force --topic {{ $topic.name }} --partitions {{ $topic.partitions }} --replication-factor {{ $topic.replicationFactor }}
+    {{- end }}
+    {{- if $topic.partitions }}
+    kafka-topics --zookeeper {{ $zk }} --alter --force --topic {{ $topic.name }} --replication-factor {{ $topic.replicationFactor }} || true
```
it looks like this command is altering replication-factor -- should be partitions

t-d-d replied:
Good catch.  Although I'm now having second thoughts about including this.  To change number of partitions for keyed topics you need to understand the implications.  I guess I can put a warning in the docs.

benjigoldberg reviewed (changes requested):
One small change

This is an interesting approach. Its reminiscent of the type of functionality you could achieve with the operator pattern except using jobs and configmaps on upgrade.  

I've been thinking about how we could add functionality to make scaling the number of nodes in the cluster up or down work, as well as handle storage issues. This is obviously well outside the scope of this PR, but perhaps an approach like this could help with some of those issues.

t-d-d commented:
@benjigoldberg This method seems to work OK for running jobs/ scripts at install/upgrade time depending on certain settings having changed.  The other use case I can see for this chart is broker runtime config settings.
I've also been thinking about scaling and rolling broker restarts (e.g. on config changes.)  I have been using the rollingUpdate.partition to apply changes over a few deployments.  I manually check whether the brokers are stable between each (check for under replicated partitions.)
I can imagine a hack with abusing the readyness probe so it only reports Ready for the current rollingUpdate.partition when underReplicatedPartition==0.  But it does feel a bit hacky.  I want to find some time to look at operators and the coreOS operator framework to see if that offers a simple solution.
What storage issues do you have in mind?

benjigoldberg commented:
@t-d-d storage issues I had in mind are essentially growing/expanding disk. With some of the changes in 1.9/1.10 and onward, it may be possible to have functionality to help with that process. Though its certainly out of scope of this PR.

t-d-d committed:
Handle sigterm in jmx exporter sidecar
